api/src/main/java/org/pmiops/workbench/notebooks/NotebooksConfig.java

```diff
@@ -37,9 +43,21 @@
   @Bean(name = USER_NOTEBOOKS_CLIENT)
   @RequestScope(proxyMode = ScopedProxyMode.DEFAULT)
   public ApiClient notebooksApiClient(
-      UserAuthentication userAuthentication, WorkbenchConfig workbenchConfig) {
+      UserAuthentication userAuthentication, WorkbenchConfig workbenchConfig, HttpHeaders headers) {
     ApiClient apiClient = buildNotebooksApiClient(workbenchConfig);
     apiClient.setAccessToken(userAuthentication.getCredentials());
+
+    // We pass-through the "Referer" header to outgoing Proxy API requests. Leonardo verifies this
+    // Header for all incoming traffic. See IA-2469, RW-6412.
+    Optional<String> referer =
+        Optional.ofNullable(headers.get("Referer"))
+            .filter(Predicates.not(List::isEmpty))
+            .map(h -> h.get(0));
+    if (referer.isPresent()) {
+      apiClient.addDefaultHeader("Referer", referer.get());
```

Comment:
in terms of the actual Referer url, i would be interested in verifying that we (the Leo team) have the correct hosts for each environment. based on testing for this PR, were you able to confirm that the Referer for dev is `all-of-us-workbench-test.appspot.com` ? we maintain our list of valid Referer headers and just want to make sure we are covering all of our bases. the other AoU hosts that we have are for prod:
"all-of-us-rw-staging.appspot.com",
"all-of-us-rw-stable.appspot.com",
"stable.fake-research-aou.org",
"preprod-workbench.researchallofus.org",
"workbench.researchallofus.org"
Let us know if you expect there to be any other host in dev or prod that will be sending requests to Leonardo :)

Reply:
It's also used for local development. With this approach, the value will include `localhost:4200` (though in theory, someone could use a different port locally) - if you can support any localhost origin, that's preferred. For the most part, those will be the only two values in dev that I can think of.

We do have a niche use case where we'd bring up a different server version with an alertnate host, e.g. for demo purposes. Often we don't need to localize files for this use case, but it would be a bit inconvenient if this stopped working. I might suggest using the list of known origins initially to verify this is working, then maybe just move to requiring *any* non-empty referer in dev.

Reply:
we include `localhost:4200` in our list of valid hosts for dev along with `localhost:3000` and `localhost:443`, we would prefer to specify ports for the local hosts that we consider valid because the Referer check is more strict that way. I'll talk to the team about eventually requiring any non-empty referer in dev. My guess is that appsec would prefer that we keep our list of valid hosts and check against that but I can ask to double check

Reply:
If appsec was fine leaving it completely off in dev, then surely this is also fine.. Seems like it should be up to the Leo team - it is a slight inconvenience to us, and might be a pain to maintain for other clients as well. Either way I would definitely go with the explicit list of valid hosts for at least the short term so we can validate the functionality with higher confidence - I'd only propose switching to a more lenient check after the feature is stable for some time.